### PR TITLE
Allow search page widget search field hint text and keyboard text input action to be customizable

### PR DESCRIPTION
**Pull request by samlythemanly**
## Description

Currently, the search page widget's hint text and keyboard text input action are hardcoded to the material localization of 'Search' and the search text input action, respectively. This prevents further customization where search functionality is desired, but the nuance of the action isn't exactly "searching" (e.g. adding a location via Google Maps Places API autocomplete).

Added a `getHintText` function (that takes in a `BuildContext`) and a `textInputAction` field to `SearchDelegate`. These default to the previous hardcoded attributes that were present in `SearchPage`.

These changes are purely aesthetic changes; no additional functionality is provided.
All tests still pass.

## Related Issues

N/A

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/search.dart [hunk 1]
```diff
@@ -89,6 +89,19 @@ Future<T> showSearch<T>({
...
+  /// button.
+  ///
+  /// Defaults to [TextInputAction.search].
+  final TextInputAction textInputAction = TextInputAction.search;
+
+  /// The hint text to be displayed in the search bar, prompting the user to
+  /// enter an input.
+  ///
+  /// Defaults to [MaterialLocalizations.searchFieldLabel].
+  String getHintText(BuildContext context) =>
```
Comment (justinmc) [inline packages/flutter/lib/src/material/search.dart]: What was your motivation for making this a function?  The alternative would be to pass it in as a String and use the localization value if it's null.  Same thing either way but I'm wondering if you have any specific use for it being a function.

**Review by justinmc — COMMENTED** [review]
@somarkoe Thanks for submitting this and sorry for the lack of reviewers.  Looks like a simple improvement to make Search a lot more useful, I'm all on board.

Just a discussion below to decide the best way to do this.  There was a Linux test failing, but I think it was flaky so I reran it.  We'll see.

**Comment by MichaelPriebe** [comment]
While you're at it, please do keyboardType: https://github.com/flutter/flutter/issues/29983
You don't always search with text.

**Comment by justinmc** [comment]
I agree with the idea to add keyboardType to this as well.

Looks like all tests are passing now.  @somarkoe are you still available to get this PR in?

**Comment by somarkoe** [comment]
Yes! Sorry, been swamped by work and haven't touched the project that deals with this in a while. I'll add the keyboard type request to it as well.

**Comment by justinmc** [comment]
Awesome, thank you @somarkoe.  Update the PR when you get a chance and I'll be happy to help you get this in.

**Comment by MichaelPriebe** [comment]
@justinmc Perhaps will you be able to get https://github.com/flutter/flutter/pull/28165 in too? There's so much about the search page that is hard coded and it shouldn't be.

**Comment by justinmc** [comment]
Yeah that one looks straightforward to me.  The original reviewer @HansMuller isn't around today but I'll talk it through with him when he's back.

**Comment by MichaelPriebe** [comment]
Abandoned? :(

File: packages/flutter/lib/src/material/search.dart (hunk 1 above)
Comment (somarkoe) [reply packages/flutter/lib/src/material/search.dart]: Either works, but I do agree, your alternative is much cleaner. Fixing!

**Comment by somarkoe** [comment]
Oops, been a while since I've used git.

Reopened and made the relevant changes. Sorry it took so long!

**Comment by MichaelPriebe** [comment]
😍 merge this ish!

**Comment by MichaelPriebe** [comment]
Only thing that could be added on is https://github.com/flutter/flutter/pull/30388 / https://github.com/flutter/flutter/pull/28165 (seems like they are the same 1 line of code)

**Comment by justinmc** [comment]
@somarkoe Thanks for following up and no worries about the delay!

Sorry to drag this out longer, but I realized that this should update the tests as well.  [This test](https://github.com/flutter/flutter/blob/master/packages/flutter/test/material/search_test.dart#L468) should probably say "keyboard shows search button by default" now.

Then, could you add a separate test that tests a different `textInputAction`?  Just duplicate that other test and add `textInputAction` to the `TestHomePage` parameters.

After that this is good to go and we can merge.

**Comment by justinmc** [comment]
@MichaelPriebe I'll try to see if we can get things moving on those 2 other PRs you referenced.

**Comment by MichaelPriebe** [comment]
@justinmc how's this going?

**Comment by justinmc** [comment]
Waiting on @somarkoe for the changes mentioned above.

**Comment by MichaelPriebe** [comment]
@somarkoe come thru

**Comment by somarkoe** [comment]
Oops, didn't notice the warning not to override, but also unsure how to alleviate this problem in the test. Thoughts?

**Comment by MichaelPriebe** [comment]
@justinmc What do you think?

**Comment by somarkoe** [comment]
I made the fields non-final since apparently overriding fields when a class is extended is no bueno, and this abstract class is one meant to be extended and not implemented (since it has functions with default behaviors).

**Comment by MichaelPriebe** [comment]
Excited for this, can finally ditch my very hacky extension of the search delegate.

**Comment by MichaelPriebe** [comment]
@justinmc Is it ready now?

**Comment by MichaelPriebe** [comment]
Lets get this pull a'mergin'!

File: packages/flutter/lib/src/material/search.dart [hunk 2]
```diff
@@ -89,6 +89,23 @@ Future<T> showSearch<T>({
 /// for another [showSearch] call.
 abstract class SearchDelegate<T> {
 
+  /// The type of action button to use for the keyboard.
```
Comment (justinmc) [inline packages/flutter/lib/src/material/search.dart]: We could keep the fields `final` by using a constructor instead like this:

```dart
  const SearchDelegate({
    this.keyboardType,
    this.hintText,
    this.textInputAction = TextInputAction.search,
  });
```

Then below, make all of the fields `final` and remove the default assignment from `textInputAction`.

File: packages/flutter/test/material/search_test.dart [hunk 3]
```diff
@@ -639,7 +653,12 @@ class _TestSearchDelegate extends SearchDelegate<String> {
     this.suggestions = 'Suggestions',
     this.result = 'Result',
     this.actions = const <Widget>[],
-  });
+    TextInputAction textInputAction,
+  }) {
+    if (textInputAction != null) {
+      this.textInputAction = textInputAction;
+    }
+  }
```
Comment (justinmc) [inline packages/flutter/test/material/search_test.dart]: Then with the change in my other comment, you can user `super` here:

```dart
...
}) : super(
  textInputAction: textInputAction,
);

**Review by justinmc — COMMENTED** [review]
Sorry for the delay!  I talked to @HansMuller to get some tips for the non-final problem.

**Comment by MichaelPriebe** [comment]
@somarkoe What do you think about the suggestions?

**Comment by MichaelPriebe** [comment]
@somarkoe Would love to hear what you think about the suggested changes.

**Comment by somarkoe** [comment]
So sorry! Was on vacation and now I'm moving 😅 

Ill take a look ASAP!

**Comment by shihaohong** [comment]
I'm going to close this since https://github.com/flutter/flutter/pull/36409 has been merged. Thank you @somarkoe for your initial work with this!

**Pull request closed without merging**